src/cohort/cohort_server.cc

```diff
@@ -1,56 +1,265 @@
+                     persist_status);
+    return;
+  }
+
+  // It's not safe to abort if there's an error here since the blockchain may
+  // have accepted our commit vote and decided to commit the transaction.
+  // TODO(benjmarks22): Add retry logic here.
+  Blockchain::Vote(request.transaction_id(), request.cohort_index(),
+                   Blockchain::Ballot::COMMIT)
+      .IgnoreError();
+  if (WaitForBlockchainDecision(request.transaction_id(),
+                                presumed_abort_time) ==
+      Blockchain::VotingDecision::COMMIT) {
+    CommitTransaction(request.transaction_id());
+  } else {
+    AbortTransaction(request.transaction_id(), request.cohort_index(),
+                     absl::nullopt);
+  }
 }
 
 grpc::Status CohortServer::PrepareTransaction(
-    ServerContext * /*context*/, const PrepareTransactionRequest *request,
-    PrepareTransactionResponse * /*response*/) {
-  results_by_transaction_id_[request->transaction_id()].config =
-      request->config();
-  results_by_transaction_id_[request->transaction_id()].future =
-      thread_pool_.submit([this, request]() { return Begin(*request); });
+    ServerContext* /*context*/, const PrepareTransactionRequest* request,
+    PrepareTransactionResponse* /*response*/) {
+  thread_pool_.push_task(
+      [this, request]() { return ProcessTransaction(*request); });
   return grpc::Status::OK;
 }
 
 grpc::Status CohortServer::GetTransactionResult(
-    ServerContext * /*context*/, const GetTransactionResultRequest *request,
-    GetTransactionResultResponse *response) {
-  const auto time =
-      FromProtoTimestamp(results_by_transaction_id_[request->transaction_id()]
-                             .config.presumed_abort_time());
-  if (results_by_transaction_id_[request->transaction_id()].future.wait_until(
-          time) == std::future_status::ready) {
-    *response =
-        results_by_transaction_id_[request->transaction_id()].future.get();
+    ServerContext* /*context*/, const GetTransactionResultRequest* request,
+    GetTransactionResultResponse* response) {
+  if (final_response_by_transaction_id_.contains(request->transaction_id())) {
+    *response = final_response_by_transaction_id_[request->transaction_id()];
+  } else {
+    response->mutable_pending_response();
   }
   return grpc::Status::OK;
 }
 
+void CohortServer::CleanUpTransactionMetadata(
+    const std::string& transaction_id) {
+  for (const auto& write_key :
+       metadata_by_transaction_id_[transaction_id].write_lock_keys) {
+    GetLock(write_key).WriterUnlock();
+  }
+  for (const auto& read_key :
+       metadata_by_transaction_id_[transaction_id].read_lock_keys) {
+    GetLock(read_key).ReaderUnlock();
+  }
+  // Should be faster than copying the response and the metadata one gets
+  // deleted right after.
+  final_response_by_transaction_id_[transaction_id].Swap(
+      &metadata_by_transaction_id_[transaction_id].response);
```

Comment:
nit: I feel this code does more than "CleanUp" which it actually loads the response for upstream to use. Maybe we pull this code out or rename the method name here.

Reply:
Renamed to ReleaseLocksAndDeleteMetadata